src/main/kotlin/com/zepben/evolve/streaming/grpc/Connect.kt

```
@@ -0,0 +1,128 @@
+    ): GrpcChannel =
+        GrpcChannelBuilder().forAddress(host, rpcPort).build()
+
+    fun connectTls(
+        host: String = "localhost",
+        rpcPort: Int = 50051,
+        ca: File? = null
+    ): GrpcChannel =
+        GrpcChannelBuilder().forAddress(host, rpcPort).makeSecure(rootCertificates = ca).build()
+
+    fun connectWithSecret(
+        clientId: String,
+        clientSecret: String,
+        confAddress: String,
+        confCAFilename: String? = null,
+        authCAFilename: String? = null,
+        host: String = "localhost",
+        rpcPort: Int = 50051,
+        ca: File? = null
+    ): GrpcChannel {
+        val tokenFetcher = createTokenFetcher(confAddress, confCAFilename = confCAFilename, authCAFilename = authCAFilename)
+            ?: return connectTls(host, rpcPort, ca)
+
+        return connectWithSecretUsingTokenFetcher(tokenFetcher, clientId, clientSecret, host, rpcPort, ca)
+    }
+
+    fun connectWithSecret(
+        clientId: String,
+        clientSecret: String,
+        audience: String,
+        issuerDomain: String,
+        authMethod: AuthMethod,
+        host: String = "localhost",
+        rpcPort: Int = 50051,
+        ca: File? = null
+    ): GrpcChannel {
+        val tokenFetcher = ZepbenTokenFetcher(audience = audience, issuerDomain = issuerDomain, authMethod = authMethod)
+
+        return connectWithSecretUsingTokenFetcher(tokenFetcher, clientId, clientSecret, host, rpcPort, ca)
+    }
+
+    fun connectWithPassword(
+        clientId: String,
+        username: String,
+        password: String,
+        confAddress: String,
+        confCAFilename: String? = null,
+        authCAFilename: String? = null,
+        host: String = "localhost",
+        rpcPort: Int = 50051,
+        ca: File? = null
+    ): GrpcChannel {
+        val tokenFetcher = createTokenFetcher(confAddress, confCAFilename = confCAFilename, authCAFilename = authCAFilename)
+            ?: return connectTls(host, rpcPort, ca)
```

@kgreav Should we throw an exception if the required auth method is none? This is like trying to unlock a door without a lock

---

It throws an exception in the python version, but I don't think it should, reason being that the intent is clear that the user wants access and the server is able to provide it.

---

Yeah you're right - It's the responsibility of the server administrators to make sure that auth is configured server side, so there's no need to alert the users if they haven't actually configured auth. As long as it always tries to use a TLS connection whenever sending credentials it should be fine - which I think it does? I.e, you can't connect to an insecure server and send it your credentials.